src/data/sparse_page_dmatrix.cc

```diff
@@ -10,9 +10,13 @@
 #if DMLC_ENABLE_STD_THREAD
 #include "./sparse_page_dmatrix.h"
 
+#include "./simple_batch_iterator.h"
+
 namespace xgboost {
 namespace data {
 
+extern template class SimpleBatchIteratorImpl<EllpackPage>;
```

Comment:
why was this required? doesn't the include in line 13 specialize this meta type with ellpackpage used later?

Reply:
This is forcing the compiler to not instantiate the template since we know it's already done in simple_dmatrix. Perhaps a bit over optimization. Removed.